source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -293,7 +302,53 @@ void SslSocket::onConnected() { ASSERT(info_->state() == Ssl::SocketState::PreHa
 
 Ssl::ConnectionInfoConstSharedPtr SslSocket::ssl() const { return info_; }
 
+void SslSocket::keylogCallback(const SSL* ssl, const char* line) {
+  if (ssl == nullptr) {
+    return;
+  }
+  BIO* bio_keylog = static_cast<BIO*> SSL_get_app_data(ssl);
+  BIO_printf(bio_keylog, "%s\n", line);
+  (void)BIO_flush(bio_keylog);
+}
+
+void SslSocket::setSSLKeyLog(bool enable) {
+  if (bio_keylog_ != nullptr) {
+    BIO_free_all(bio_keylog_);
+    bio_keylog_ = nullptr;
+  }
+
+  SSL_CTX* ctx = SSL_get_SSL_CTX(rawSsl());
+  if (ctx == nullptr || tls_key_log_path_ == nullptr) {
+    return;
+  }
+  if (!enable) {
+    ENVOY_LOG(debug, "Disable ssl key log");
+    SSL_CTX_set_keylog_callback(ctx, nullptr);
+    enable_tlskey_log_ = false;
+    return;
+  }
+  /*
+   * Append rather than write in order to allow concurrent modification.
+   * Furthermore, this preserves existing key log files which is useful when
+   * the tool is run multiple times.
+   */
+  bio_keylog_ = BIO_new_file(tls_key_log_path_->c_str(), "a");
+  if (bio_keylog_ == nullptr) {
+    return;
+  }
+
+  ENVOY_LOG(debug, "Enable ssl key log");
+  SSL* ssl = rawSsl();
+  SSL_set_app_data(ssl, bio_keylog_);
```

Comment:
SSL app data could be called at some other places, e.g. https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/listener/tls_inspector/tls_inspector.cc#L83

Is it necessary to check if app data already set before call `SSL_set_app_data()` here?

Reply:
I will refactor it.